Add Handlers

jacobweinstock opened the pull request:
## Description

<!--- Please describe what this PR is going to change -->
This PR adds handler functions per message type and allows us to participate in the full DORA(discover, offer, request, ack) cycle. This PR is built on https://github.com/tinkerbell/dhcp/pull/2 and requires it to be merged before this one. 

## Why is this needed

<!--- Link to issue you have raised -->

Fixes: #

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->


## How are existing users impacted? What migration steps/scripts do we need?

<!--- Fixes a bug, unblocks installation, removes a component of the stack etc -->
<!--- Requires a DB migration script, etc. -->


## Checklist:

I have:

- [x] updated the documentation and/or roadmap (if required)
- [x] added unit or e2e tests
- [x] provided instructions on how to upgrade

jacobweinstock committed:
Add handler functions per message type:

This allows us to participate in the full
DORA(discover,offer,request,ack) cycle.
The release handler is a noop but include for
clarity. Release is a noop because the design of
this repo is that all DHCP addresses are reservations
so a release doesn't put an IP back into a pool.

Signed-off-by: Jacob Weinstock <jakobweinstock@gmail.com>

jacobweinstock committed:
Add open telemetry tracing:

This gives us some basic otel plumbing included.

Signed-off-by: Jacob Weinstock <jakobweinstock@gmail.com>

jacobweinstock committed:
Update and add tests for handlers:

Test bad WriteTo call.

Signed-off-by: Jacob Weinstock <jakobweinstock@gmail.com>

jacobweinstock committed:
Add method to determine if a packet is from a netboot enabled client:

This allows us to filter out DHCP only clients from clients
requesting to netboot.

Signed-off-by: Jacob Weinstock <jakobweinstock@gmail.com>

jacobweinstock committed:
Add logging:

Help with troubleshooting netboot client issues.

Signed-off-by: Jacob Weinstock <jakobweinstock@gmail.com>

nshalman commented on handler.go:
```diff
@@ -0,0 +1,214 @@
+		trace.WithAttributes(attribute.String("MessageType", m.MessageType().String())),
+	)
+	defer span.End()
+	var reply *dhcpv4.DHCPv4
+	switch mt := m.MessageType(); mt {
+	case dhcpv4.MessageTypeDiscover:
+		reply = s.handleDiscover(ctx, tracer, m)
+	case dhcpv4.MessageTypeRequest:
+		reply = s.handleRequest(ctx, tracer, m)
+	case dhcpv4.MessageTypeRelease:
+		s.handleRelease(ctx, m)
+	default:
+		log.Info("received unknown message type", "type", mt)
+	}
+	c := codes.Ok
+	cmsg := fmt.Sprintf("no reply send DHCP: %v", m.MessageType())
+	if reply != nil {
```
I don't see logic updating the span to reflect that the reply was unexpectedly nil.
That seems like a case we might care about.

tobert replied:
in boots I handled this by calling span.SetStatus in the error handling path

nshalman commented on handler.go:
```diff
@@ -0,0 +1,214 @@
+		attribute.Bool("AllowNetboot", n.AllowNetboot),
+		attribute.String("IpxeScriptURL", fmt.Sprintf("%v", n.IpxeScriptURL)),
+	)
+	span.SetStatus(codes.Ok, "done reading from backend")
+	span.End()
```
Perhaps `span.End()` should be called with `defer`?
Also, the "done reading from backend" doesn't quite seem to match "DHCP Discover Message".
I think the tracing in this function could be a little better.

tobert replied:
yep, in Go, you typically want to defer span.End unless there's some reason you'd want to drop the span (bc span.End is when they actually get sent)

nshalman commented on handler.go:
```diff
@@ -0,0 +1,214 @@
+	ctx, span := tracer.Start(ctx, "DHCP Request Message")
+	log.Info("received request packet")
+	d, n, err := s.Backend.Read(ctx, m.ClientHWAddr)
+	if err != nil {
```
In this code path I don't think span.End() gets called at all.
I think the pattern of using defer would be preferable.

jacobweinstock replied:
my understanding of otel, admittedly very shallow, is that we start and end spans around functionality we want to be timed. I don't know if that is correct, but that is the intention here. `s.Backend.Read` will typically contain some kind of network call. Moving to a `defer span.End()` would change the timing to include more than just the underlining network call of `s.Backend.Read`. Is that ok? or "the way" to do otel? I'm definitely open to changing this. what do you think?

nshalman replied:
If we're **only** concerned with timing Backend.Read, perhaps we should explicitly trace just that, e.g.

```go
	ctx, span := tracer.Start(ctx, "querying backend")
	d, n, err := s.Backend.Read(ctx, m.ClientHWAddr)
        span.End()
```

The current code gives mixed messages. It looks like it's tracing the "DHCP Request Message".
Also, couldn't we just do tracing within the Backend using the provided context?
Could be worth setting up a testbed to look at what the traces look like to see if they look informative?
Happy to discuss further.

nshalman reviewed (changes requested):
(As I should have said on #2) Thank you for all this great work!
I think there's a bug in the tracing in one of the functions that should be fixed.
My suggestion is to use `defer` on `span.End()`.

jacobweinstock committed:
Add comment to DHCP data type:

Clarify that this type is the API
between the DHCP server and any backends.

Signed-off-by: Jacob Weinstock <jakobweinstock@gmail.com>

jacobweinstock commented:
Hey @nshalman, @tobert, I refactored a bit. mind taking a look?

nshalman commented on handler.go:
```diff
@@ -0,0 +1,192 @@
+	return d, n, nil
+}
+
+// handleDiscover handles DHCP packets message type of discover.
```
comment no longer matches function name

jacobweinstock replied:
ah thanks. updated.

nshalman reviewed (changes requested):
One nit on a comment that fell out of sync with the function it documents, but otherwise, this looks great to me! Thank you!

jacobweinstock committed:
Update handler funcs:

Tried to reduce the cyclomatic complexity of handlerFunc.
refactors functions to share a little more code so that
adding otel was less cumbersome.

Signed-off-by: Jacob Weinstock <jakobweinstock@gmail.com>

nshalman reviewed (approved):
Nice work!
